docs/geps/gep-04.rst

```diff
@@ -0,0 +1,316 @@
+in the taxes and transfers system are evaluated. This ensures that the inputs are
+already computed before a function that requires them is called.
+
+In order to calculate a set of taxes and transfers, GETTSIM builds a DAG based on three
+inputs provided by the user:
+
+ - Input data.
+ - A set of functions representing the tax and transfer system, which consist of the
+   ones pre-implemented in GETTSIM and potentially user-written additional functions.
+   Parameters of the taxes and transfers system (see :ref:`gep-3` will already be
+   partialled into these functions, so they can be ignored in the following).
+ - The target columns of interest.
+
+The DAG is then used to call all required functions in the right order and to calculate
+the requested targets.
+
+
+Level of the DAG and limitations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In principle, GETTSIM will import all functions defined in the modules describing the
+taxes and transfers system. In principle, these functions refer to all years in
+GETTSIM's scope. There has to be some discretion in order to allow for the interface of
+functions to change over time, new functions to appear, or old ones to disappear.
+
+Some examples include:
+
+1. ``arbeitsl_hilfe`` being replaced by ``arbeitsl_geld_2``
+1. ``kinderbonus`` being active only in a few years
+1. The introduction of ``kinderzuschlag``
+1. Capital income entering ``sum_brutto_eink`` or not.
+
+The goal is that the graph for any particular point in time is minimal in the sense that
+``arbeitsl_geld_2`` does not appear before it was conceived, it is apparent from the
+interface of ``sum_brutto_eink`` whether it includes capital income or not, etc..
+
+In the yaml-files corresponding to a particular tax / transfer, functions not present in
+all years will need to be listed with along with the dates for when they are active. See
+:gep-3-keys-referring-to-functions: for the precise syntax. That mechanism should be
+used for:
+
+1. Functions that are newly introduced.
+2. Functions that cease to be relevant.
+3. Functions whose interface changes over time.
+4. Functions whose body changes so much that
+
+   - it is useful to signal that things have changed and/or
+
+   - it would be awkward to program the different behaviors in one block with case
+     distinctions.
+
+Needless to say, the different reasons may appear at different points in time for the
+same function.
+
+
+Additional functionalities
+--------------------------
+
+We implemented a small set of additional features that simplify the specification of
+certain types of functions of the tax and transfer system.
+
+
+.. _gep-4-group-sums:
+
+Group summation of variables
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Individual-level measures often need to be summed up to a group level. <GEP-2 describes
+`gep-2-aggregation-functions`> how reductions are handled in principle.
+
+Because summations are very common, GETTSIM's DAG backend automates this process for
```

Comment:
We need to describe somewhere how users can aggregate functions. This won't be straightforward once we move to scalar. But probably best to wait until we know more about the implementation.

Reply:
Yes, in GEP-2. The only thing relevant for the DAG is the automatic conversion because it injects nodes into the DAG. What other nodes do precisely, does not matter from the DAG's perspective (it will continue to work on columns, same behaviour as we have currently).